[Trisha Patodia] iP

triishaa opened the pull request:
# ChatterBox : Enhancing User Interaction 💬 

> "For every minute spent in organizing, an hour is earned." - Benjamin Franklin ([source](https://www.teachingcove.com/motivate/every-minute-spent-organizing/))

#### ChatterBox is a personal assistant chatbot designed to help you stay organized by managing tasks like ToDos, Deadlines, and Events. This update focuses on improving user experience and adding new functionalities. 

### Features Implemented: 
 - Add Tasks (`ToDo`, `Deadline`, and `Event`) 
 - Mark tasks as done ✅  or not done ❌ 
 - Save  and load tasks from file
 - Print tasks which all have a keyword 🔍 
 - Improve error handling for user inputs

 ### Next Steps:
1. **Refactor**  the code for better readability and maintainability.
2. **Optimise** file _I/O_ operations to enhance performance. 
3. ~~Add~~ **Revise** unit tests to cover edge cases and improve readability.

### Task List: 
- [x] Create a basic chatbot framework. 
- [x] Implement `add` , `delete` , `mark` and `search` functionalities.
- [ ] Improve user interface for better user experience. 

### **_Example Code Snippet_**
#### Here's a code snipped explaining how the `mark` functionality is implemented: 
```java 
private void markCommand(String input, TaskList taskList, Storage storage, Ui ui) {
        try {
            int indexSpace = input.indexOf(" ");
            int taskIndex = Integer.parseInt(input.substring(indexSpace + 1)) - 1;
            if (taskIndex >= 0 && taskIndex < taskList.size()) {
                taskList.getTask(taskIndex).markAsDone();
                storage.saveTasks(taskList.getTasks());
                ui.showTaskMarkedAsDone(taskList.getTask(taskIndex));
            } else {
                ui.showErrorInvalidTaskNumber();
            }
        } catch (NumberFormatException e) {
            ui.showError("Command must be followed by a specific task number");
        }
    }
```


### For more information, visit the [ChatterBox GitHub Repository](https://github.com/triishaa/ip)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Renamed Duke.java to ChatterBox.java and added initial functionality

Commit:
Level 1 Echo: Implemented echo functionality and exit command

Commit:
Added supoort for list and personalised the chatbot

Commit:
Level 2 Add, List: Implemented the functionality to add tasks to list and also view them.

Commit:
Added Level 3: Mark and Unmark tasks functionality

triishaa committed:
Level-3 updated

Commit:
Level-3 Final, Used Task Array instead of an Array List for mark and unmark task functionality

Commit:
Resolved merge conflicts

Commit:
Added Level 4: ToDos, Events, Deadlines Functionality by creating three subclasses of class Task

Commit:
A-TextUiTesting Done

Commit:
Add Level-5: Handle Errors which checks for incorrect commands, blank description, incorrect format of entering tasks,invalid task numbers to be marked done/undone

Commit:
Add Level 6: Deleting tasks functionality added

Commit:
Add Level 7: Add taks to file

Commit:
Merge branch 'branch-Level-7'

Merge branch 'branch-Level-7' into master

Implemented Level 7:
- Enhanced ChatterBox to keep track of previous entries using a text file.

Commit:
Add Level 8: ChatBot now recognises date and time

Commit:
Merge branch-Leel-8 into master

Commit:
Add A-MoreOO
P: Divided the main ChatterBox class into sub-classes which handle similar functions

Commit:
A-MoreOOP : ChatterBox is now divided into subclasses where each subclass performs a specific task.

Commit:
Add A-Packages: Classes are now grouped into packages which perform similar task

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Merged origin/add-graddle-support into master

Commit:
Add A-Gradle: Set up gradle in IntelliJ

Commit:
Add JUnit to test Parser, Ui and Storage Classes

Commit:
Add Javadoc to the project

Commit:
Implement coding standard

Commit:
Add Level 9: Find feature on Chatbot

k-hian commented on src/main/java/command/Parser.java:
```diff
@@ -0,0 +1,240 @@
+     * @param ui The user interface handler for displaying messages.
+     */
+    public void parseExecute(String input, TaskList taskList, Storage storage, Ui ui) {
+        if(input.startsWith("mark")) {
```
Take note on single space within the statements. Here, there should be a space between `if` and the condition.

k-hian commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,79 @@
+ * A Deadline task has a description and a specific date and time
+ * by which it needs to be completed.
+ */
+public class Deadline extends Task{
```
Take note of the missing single space between `Task` and `{`.

Similarly for a few other classes as well.

k-hian commented on src/test/java/command/UiTest.java:
```diff
@@ -0,0 +1,66 @@
+    private final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+    private final PrintStream originalOut = System.out;
+
+    @BeforeEach
```
First time seeing a setup and teardown method. Learned something new :+1:

k-hian reviewed (commented):
LGTM. Just some minor nits to fix.

Commit:
Add CheckStyle

Commit:
Add Level 10: GUI Implementation

nihalramesh commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,79 @@
+     * @return The description of the deadline task.
+     */
+    @Override
+    public String getDescription() {
```
Perhaps we could get rid of this method, as it adds no functionality as compared to its parent method?

nihalramesh commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,93 @@
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: "
+                + dateStart.format(DateTimeFormatter.ofPattern("MMM dd yyyy hh:mm a"))
```
To avoid repeated code and ensure consistency, consider defining a DateTimeFormatter as a private static final constant for the two formats used in toString() and toSaveFormat().

nihalramesh commented on src/main/java/command/Parser.java:
```diff
@@ -0,0 +1,237 @@
+     * @param storage The storage handler responsible for saving tasks.
+     * @param ui The user interface handler for displaying messages.
+     */
+    public void parseExecute(String input, TaskList taskList, Storage storage, Ui ui) {
```
Perhaps we could rename this to something more descriptive?

nihalramesh commented on src/main/java/command/Parser.java:
```diff
@@ -0,0 +1,237 @@
+     */
+    private void deadlineCommand(String input, TaskList taskList, Storage storage, Ui ui) {
+        if (!input.contains("/by")) {
+            ui.showError("task.Deadline format should be: deadline DESCRIPTION /by DATE");
```
Perhaps we could standardised error messages?

nihalramesh reviewed (commented):
Overall LGTM, just a bit of nit picking here and there. Cheers!

Commit:
Add assertions to improve code reliability

-Added assertions to verify preconditions and postconditions in key methods.
-Included assertions to check for invalid inputs.
-Refactored code to ensure that the assertions are properly checked during execution.

This change enhances code reliability by ensuring that assumptions are validated during runtime.

Commit:
Improve Code Quality by Enhancing Readability and Structure

Extracted methods to reduce the size of previously existing methods, making them more manageable and focused on a single responsibility.
Removed unused variables and methods, streamlining the code and improving clarity.
Applied KISS (Keep It Simple, Stupid) principles to simplify complex code, making it easier to understand and maintain.
Reorganized code logically to enhance readability and flow, ensuring a more intuitive structure.
Utilized the default branch for code consistency and ease of management.
Avoided recycling variables and parameters to prevent unintended side effects and increase code clarity.
Removed empty catch blocks to handle exceptions properly and avoid silent failures.
Deleted dead code to eliminate unnecessary clutter and improve performance.
Added comments to clarify the purpose and functionality of code sections, aiding in understanding and future maintenance.

These changes collectively enhance the overall quality of the code, making it more readable, maintainable, and robust.

Commit:
Improve Code Quality by Enhancing Readability and Structure

- Extracted methods to reduce the size of previously existing methods, making them more manageable and focused on a single responsibility.
- Removed unused variables and methods, streamlining the code and improving clarity.
- Applied KISS (Keep It Simple, Stupid) principles to simplify complex code, making it easier to understand and maintain.
- Reorganized code logically to enhance readability and flow, ensuring a more intuitive structure.
- Utilized the default branch for code consistency and ease of management.
- Avoided recycling variables and parameters to prevent unintended side effects and increase code clarity.
- Removed empty catch blocks to handle exceptions properly and avoid silent failures.
- Deleted dead code to eliminate unnecessary clutter and improve performance.
- Added comments to clarify the purpose and functionality of code sections, aiding in understanding and future maintenance.

These changes collectively enhance the overall quality of the code, making it more readable, maintainable, and robust.

Commit:
Add assertions to improve code reliability

- Added assertions to verify preconditions and postconditions in key methods.
- Included assertions to check for invalid inputs.
- Refactored code to ensure that assertions are properly checked during execution.

This change enhances code reliability by ensuring that assumptions are validated during runtime.

Commit:
Refactor findTasks Method to use Streams

Replaced Iteration with Streams: Updated the findTasks method in the TaskList class to utilize Java Streams for more concise and expressive filtering.

Filtering Logic: Leveraged Stream API's filter method to search for tasks with descriptions that contain the specified search term, ignoring case.

Result Collection: Collected the filtered results into a List<Task> and converted it back to an ArrayList<Task> for compatibility with existing code.

Edge Case Handling: Ensured that if the provided description is null or empty, the method returns an empty ArrayList to handle such scenarios gracefully.

triishaa committed:
Update Parser.java

triishaa committed:
Merge pull request #1 from triishaa/branch-A-Assertions

Add Input Validation and Error Handling

triishaa committed:
Merge branch 'master' into branch-A-CodeQuality

triishaa committed:
Merge pull request #2 from triishaa/branch-A-CodeQuality

Branch a code quality

triishaa committed:
Merge branch 'master' into branch-A-Streams

triishaa committed:
Merge pull request #3 from triishaa/branch-A-Streams

Refactor findTasks Method to use Streams

Commit:
Add D-Merchandise

Commit:
Add D-Merchandise Extension

Commit:
Merge branch 'branch-A-DMerchandise'

Merge branch

Commit:
Style: update text font and error message styling

Commit:
Add representative screenshot of product

Commit:
Add representative screenshot of product

Commit:
Update README with new feature descriptions
